llama_index/vector_stores/postgres.py

```diff
@@ -0,0 +1,158 @@
+from typing import List, Any, Type, Optional
+
+from llama_index.data_structs.node import DocumentRelationship, Node
+from llama_index.vector_stores.types import (
+    VectorStore,
+    NodeWithEmbedding,
+    VectorStoreQuery,
+    VectorStoreQueryResult,
+)
+
+
+def get_data_model(base: Type, index_name: str) -> Any:
+    """
+    This part create a dynamic sqlalchemy model with a new table
+    """
+    from pgvector.sqlalchemy import Vector
+    from sqlalchemy import Column
+    from sqlalchemy.dialects.postgresql import BIGINT, VARCHAR, JSON
+
+    class AbstractData(base):  # type: ignore
+        __abstract__ = True  # tShis line is necessary
+        id = Column(BIGINT, primary_key=True, autoincrement=True)
+        text = Column(VARCHAR, nullable=False)
+        extra_info = Column(JSON)
+        doc_id = Column(VARCHAR)
+        embedding = Column(Vector(1536))  # type: ignore
+
+    tablename = "data_%s" % index_name  # dynamic table name
+    class_name = "Data%s" % index_name  # dynamic class name
+    model = type(class_name, (AbstractData,), {"__tablename__": tablename})
+    return model
+
+
+class PGVectorStore(VectorStore):
+    stores_text = True
+
+    def __init__(self, connection_string: str, table_name: str) -> None:
+        try:
+            import sqlalchemy  # noqa: F401
+        except ImportError:
+            raise ImportError("`sqlalchemy` package not found")
+
+        try:
+            import pgvector  # noqa: F401
+        except ImportError:
+            raise ImportError("`pgvector` package not found")
+
+        try:
+            import psycopg2  # noqa: F401
+        except ImportError:
+            raise ImportError("`psycopg2-binary` package not found")
+
+        from sqlalchemy.orm import declarative_base
+
+        self.connection_string = connection_string
+        self.table_name: str = table_name.lower()
+        self._conn = self._connect()
+        self._base = declarative_base()
+        # sqlalchemy model
+        self.table_class = get_data_model(self._base, self.table_name)
+        self._create_extension()
+        self._create_tables_if_not_exists()
+
+    @classmethod
+    def conn_str_from_params(
```

Comment:
nit: I think easier to just have a `def from_params` that takes in the params you have here and table_name, and return the VectorStore directly instead of the connection string.

Reply:
done